### PR TITLE
`py-wandb`: add v0.13.9

### PR DESCRIPTION
**Pull request by thomas-bouvier**
Quite many changes in dependencies

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 1]
```diff
@@ -10,23 +10,48 @@ class PyWandb(PythonPackage):
...
 
-    homepage = "https://github.com/wandb/"
-    url = "https://github.com/wandb/client/archive/v0.10.1.tar.gz"
+    homepage = "https://github.com/wandb/wandb"
+    pypi = "wandb/wandb-0.13.9.tar.gz"
 
+    maintainers = ["thomas-bouvier"]
+
+    version("0.13.9", sha256="0a17365ce1f18306ce7a7f16b943094fac7284bb85f4e52c0685705602f9e307")
     version("0.10.1", sha256="abd334cd1460ac1f6e5aa959d3e04c46cd246f96cfc3323fc0572916760d32ab")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]:
```suggestion
    version("0.10.1", sha256="d02427cda58a6618ba10a027a76d9e3f68ad923d35964b1b68785c49e5160009", deprecated=True)
```
PyPI tarballs have different checksums than GitHub. Also, the 0.10.1 release was yanked from PyPI suggesting it contains a serious bug, so we should probably deprecate it.

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 2]
```diff
@@ -10,23 +10,48 @@ class PyWandb(PythonPackage):
...
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]: If it makes it easier to manage deps, you can also do:
```python
with when("@0.13.9"):
    depends_on(...)

with when("@0.10.1"):
    depends_on(...)
```
Then you don't need a `when="..."` for every `depends_on`

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 3]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
+        depends_on("py-nvidia-ml-py3", type=("build", "run"))
+        depends_on("py-python-dateutil", type=("build", "run"))
+        depends_on("py-shortuuid", type=("build", "run"))
+        depends_on("py-six", type=("build", "run"))
+        depends_on("py-watchdog", type=("build", "run"))
+        depends_on("py-configparser", type=("build", "run"))
+
+    depends_on("py-pyyaml", type=("build", "run"))
     depends_on("py-click", type=("build", "run"))
+    depends_on("py-click@7.0:", type=("build", "run"), when="@0.13.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]:
```suggestion
    depends_on("py-click@7,8.0.1:", type=("build", "run"), when="@0.13.0:")
```
or:
```suggestion
    depends_on("py-click@7:", type=("build", "run"), when="@0.13.0:")
    conflicts("^py-click@8.0.0")
```

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 4]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
-    depends_on("py-gql", type=("build", "run"))
-    depends_on("py-nvidia-ml-py3", type=("build", "run"))
-    depends_on("py-python-dateutil", type=("build", "run"))
+    depends_on("py-gitpython@1.0.0:", type=("build", "run"), when="@0.13.0:")
     depends_on("py-requests", type=("build", "run"))
-    depends_on("py-shortuuid", type=("build", "run"))
-    depends_on("py-six", type=("build", "run"))
-    depends_on("py-watchdog", type=("build", "run"))
-    depends_on("py-pyyaml", type=("build", "run"))
+    depends_on("py-requests@2.0.0:", type=("build", "run"), when="@0.13.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]:
```suggestion
    depends_on("py-requests@2", type=("build", "run"), when="@0.13.0:")
```

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 5]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
+    with when("@0.13.0:"):
+        depends_on("py-grpcio@1.46.3:", type="build")
+        depends_on("py-grpcio-tools@1.46.3:", type="build")
+        depends_on("py-twine", type="build")
+        depends_on("py-pathtools", type=("build", "run"))
+        depends_on("py-setproctitle", type=("build", "run"))
+        depends_on("py-appdirs@1.4.3:", type=("build", "run"))
+        depends_on("py-protobuf@3.12.0:4", type=("build", "run"), when="^python@:3.8 platform=linux")
+        depends_on("py-protobuf@3.15.0:4", type=("build", "run"), when="^python@3.9 platform=linux")
+        depends_on("py-protobuf@3.19.0:4", type=("build", "run"), when="^python@3.10: platform=linux")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]: You can add:
```python
conflicts("^py-protobuf@4.21.0")
```
to handle the `!=`, or add it to the version range in `depends_on`: `3.12:4.20,4.21.1:4`

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 6]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
 
+    version("0.13.9", sha256="0a17365ce1f18306ce7a7f16b943094fac7284bb85f4e52c0685705602f9e307")
+    version("0.10.1", sha256="d02427cda58a6618ba10a027a76d9e3f68ad923d35964b1b68785c49e5160009", deprecated=True)
+
+    depends_on("python@3.6:", type=("build", "link", "run"), when="@0.13.0:")
     depends_on("py-setuptools", type="build")
+
+    with when("@0.13.0:"):
+        depends_on("py-grpcio@1.46.3:", type="build")
+        depends_on("py-grpcio-tools@1.46.3:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]: These are optional, not required

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 7]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
+        depends_on("py-twine", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]: Not a dependency

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 8]
```diff
@@ -10,23 +10,49 @@ class PyWandb(PythonPackage):
...
+    pypi = "wandb/wandb-0.13.9.tar.gz"
 
-    version("0.10.1", sha256="abd334cd1460ac1f6e5aa959d3e04c46cd246f96cfc3323fc0572916760d32ab")
+    maintainers = ["thomas-bouvier"]
 
+    version("0.13.9", sha256="0a17365ce1f18306ce7a7f16b943094fac7284bb85f4e52c0685705602f9e307")
+    version("0.10.1", sha256="d02427cda58a6618ba10a027a76d9e3f68ad923d35964b1b68785c49e5160009", deprecated=True)
+
+    depends_on("python@3.6:", type=("build", "link", "run"), when="@0.13.0:")
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]:
```suggestion
    depends_on("py-setuptools", type=("build", "run"))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-wandb/package.py (hunk 8 above)
Comment (thomas-bouvier) [reply var/spack/repos/builtin/packages/py-wandb/package.py]: Should `py-setuptools` always have the `run` dependency?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-wandb/package.py]: Nope, rarely. In this case, `pkg_resources` is imported at run-time. And it's listed under `install_requires`, not `setup_requires`.

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 9]
```diff
@@ -10,23 +10,48 @@ class PyWandb(PythonPackage):
...
-    depends_on("py-gql", type=("build", "run"))
-    depends_on("py-nvidia-ml-py3", type=("build", "run"))
-    depends_on("py-python-dateutil", type=("build", "run"))
+    depends_on("py-gitpython@1:", type=("build", "run"), when="@0.13.0:")
     depends_on("py-requests", type=("build", "run"))
-    depends_on("py-shortuuid", type=("build", "run"))
-    depends_on("py-six", type=("build", "run"))
-    depends_on("py-watchdog", type=("build", "run"))
-    depends_on("py-pyyaml", type=("build", "run"))
+    depends_on("py-requests@2:", type=("build", "run"), when="@0.13.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]:
```suggestion
    depends_on("py-requests@2", type=("build", "run"), when="@0.13.0:")
```

File: var/spack/repos/builtin/packages/py-wandb/package.py [hunk 10]
```diff
@@ -10,23 +10,48 @@ class PyWandb(PythonPackage):
...
+    depends_on("python@3.6:", type=("build", "link", "run"), when="@0.13.0:")
+    depends_on("py-setuptools", type=("build", "run"))
+
+    with when("@0.13.0:"):
+        depends_on("py-pathtools", type=("build", "run"))
+        depends_on("py-setproctitle", type=("build", "run"))
+        depends_on("py-appdirs@1.4.3:", type=("build", "run"))
+        depends_on("py-protobuf@3.12.0:4", type=("build", "run"), when="^python@:3.8 platform=linux")
+        depends_on("py-protobuf@3.15.0:4", type=("build", "run"), when="^python@3.9 platform=linux")
+        depends_on("py-protobuf@3.19.0:4", type=("build", "run"), when="^python@3.10: platform=linux")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-wandb/package.py]: Also needed for non-linux platforms. Honestly I would probably just do:
```python
depends_on("py-protobuf@3.19:4", type=("build", "run"))
```
for all versions/OSes and skip the complexity

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-wandb/package.py (hunk 10 above)
Comment (thomas-bouvier) [reply var/spack/repos/builtin/packages/py-wandb/package.py]: Alright let's do this :)

**Comment by adamjstewart** [comment]
@spackbot fix style

**Review by adamjstewart — DISMISSED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**